core_matcher.go

```diff
@@ -19,34 +19,48 @@ import (
 )
 
 // coreMatcher uses a finite automaton to implement the matchesForJSONEvent and MatchesForFields functions.
-// state is the start of the automaton
-// namesUsed is a map of field names that are used in any of the patterns that this automaton encodes. Typically,
-// patterns only consider a subset of the fields in an incoming data object, and there is no reason to consider
-// fields that do not appear in patterns when using the automaton for matching
-// the updateable fields are grouped into the coreStart member so they can be updated atomically using atomic.Load()
+// The updateable fields are grouped into the coreFields member so they can be updated atomically using atomic.Load()
 // and atomic.Store(). This is necessary for coreMatcher to be thread-safe.
 type coreMatcher struct {
-	updateable atomic.Value // always holds a *coreStart
+	updateable atomic.Value // always holds a *coreFields
 	lock       sync.Mutex
 }
-type coreStart struct {
-	state                     *fieldMatcher
-	namesUsed                 map[string]bool
-	presumedExistFalseMatches *matchSet
+
+// coreFields groups the updateable fields in coreMatcher.
+// state is the start of the automaton.
+// namesUsed is a map of field names that are used in any of the patterns that this automaton encodes. Typically,
+// patterns only consider a subset of the fields in an incoming data object, and there is no reason to consider
+// fields that do not appear in patterns when using the automaton for matching.
+// fakeField is used when the flattener for an event returns no fields, because it could still match if
+// there were patterns with "exists":false. So in this case we run one fake field through the matcher
+// which will cause it to notice that any "exists":false patterns should match.
+type coreFields struct {
+	state     *fieldMatcher
+	namesUsed map[string]bool
+	fakeField []Field
```

Comment:
From what I can tell how this is used this can be simplified:

- drop this field
- create and use a global variable instead

Wondering if we need this rather "static" object in the struct and instead make it a global variable?

```var fakeField = Field{
 		Path:       []byte{byte(byteCeiling)},
 		Val:        []byte(""),
 		ArrayTrail: []ArrayPos{{0, 0}},
 	}
```

```
	if len(fields) == 0 {
		fields = []Field(fakeField)
	}
```

Reply:
Or even simpler (without needing the global variable):

```
	if len(fields) == 0 {
		fields = []Field{<construct inline fake field>}
	}
```

Reply:
Yes, that's the right thing. Earlier, I thought that the fakeField would be required in several places, so I stored it in the coreMatcher object, since there's only one coreMatcher. I suppose a global variable would be better but this is even better.